gpbft/gpbft_test.go

```diff
@@ -0,0 +1,457 @@
+package gpbft_test
+
+import (
+	"testing"
+
+	"github.com/filecoin-project/go-f3/gpbft"
+	"github.com/filecoin-project/go-f3/gpbft/emulator"
+)
+
+var (
+	tipset0 = gpbft.TipSet{Epoch: 0, Key: []byte("bigbang")}
+	tipSet1 = gpbft.TipSet{Epoch: 1, Key: []byte("fish")}
+	tipSet2 = gpbft.TipSet{Epoch: 2, Key: []byte("lobster")}
+)
+
+func TestGPBFT_DecidesProposalOnStrongEvidenceOfPrepare(t *testing.T) {
+	require := emulator.NewAssertions(t)
+
+	instance, err := emulator.NewInstance(
+		0,
+		gpbft.PowerEntries{
+			gpbft.PowerEntry{
+				ID:    0,
+				Power: gpbft.NewStoragePower(1),
+			},
+			gpbft.PowerEntry{
+				ID:    1,
+				Power: gpbft.NewStoragePower(1),
+			},
+		},
+		tipset0, tipSet1, tipSet2,
+	)
+	require.NoError(err)
+
+	em, err := emulator.New()
+	require.NoError(err)
+
+	// Start the instance and trigger the first alarm because the GPBFT
+	// implementation in this repo implicitly uses the alarm to actually start the
+	// instance.
+	require.NoError(em.Start(instance))
+	require.AlarmReceived(em.ReceiveAlarm())
+
+	// Expect QUALITY step from the emulator.
+	require.Quality(require.MessageReceived(em.ReceiveBroadcast()), instance)
+
+	// Send QUALITY to the emulator to satisfy strong quorum of senders.
+	require.MessageReceived(
+		em.ReceiveMessage(1, instance.NewQualityForProposal()),
+	)
+
+	// Expect PREPARE step with the original instance proposal, since strong quorum
+	// of senders is met during QUALITY phase.
+	receivedPrepare := require.MessageReceived(em.ReceiveBroadcast())
+	require.Prepare(receivedPrepare, instance)
+
+	// Send PREPARE message to the emulator for the instance proposal which should
+	// facilitate evidence of strong quorum of PREPARE.
+	sentPrepare := require.MessageReceived(
```

Comment:
Received/sent names are confusing here. receivedPrepare is the on the participant sent, not the one it received from someone else (I think).

Reply:
Thanks for pointing that out. I'll see what can do. 

The way this makes sense in my head is that the writer of the test is the "subject" of `sent` `received` naming prefix of the variables. For example, `sentPrepare` means the test sent a prepare message. The test can never send a message on behalf of the participant. It can receive it (by delivering a self-broadcast). Hence the `recievedPrepare` naming.